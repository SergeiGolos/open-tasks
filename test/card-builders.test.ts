import { describe, it, expect } from 'vitest';
import {
  createCardBuilder,
  QuietCardBuilder,
  SummaryCardBuilder,
  VerboseCardBuilder,
} from '../src/card-builders.js';
import type { CardContent, CardStyle } from '../src/types.js';

describe('createCardBuilder factory', () => {
  it('should create QuietCardBuilder for quiet level', () => {
    const builder = createCardBuilder('quiet');
    expect(builder).toBeInstanceOf(QuietCardBuilder);
  });

  it('should create SummaryCardBuilder for summary level', () => {
    const builder = createCardBuilder('summary');
    expect(builder).toBeInstanceOf(SummaryCardBuilder);
  });

  it('should create VerboseCardBuilder for verbose level', () => {
    const builder = createCardBuilder('verbose');
    expect(builder).toBeInstanceOf(VerboseCardBuilder);
  });
});

describe('VerboseCardBuilder', () => {
  it('should render a table card with borders', () => {
    const builder = new VerboseCardBuilder();
    const tableContent: CardContent = {
      type: 'table',
      headers: ['Name', 'Status'],
      rows: [
        ['Task 1', 'Success'],
        ['Task 2', 'Failed'],
      ],
    };
    builder.addCard('Test Table', tableContent, 'info');
    const output = builder.build();
    expect(output).toContain('Test Table');
    expect(output).toContain('Name');
    expect(output).toContain('Status');
    expect(output).toContain('Task 1');
    expect(output).toContain('Success');
    expect(output).toContain('Task 2');
    expect(output).toContain('Failed');
    // Check for boxen characters
    expect(output).toMatch(/╭.*╮/);
    expect(output).toMatch(/╰.*╯/);
  });

  it('should render a list card with borders', () => {
    const builder = new VerboseCardBuilder();
    const listContent: CardContent = {
      type: 'list',
      items: ['Item 1', 'Item 2'],
    };
    builder.addCard('Test List', listContent, 'success');
    const output = builder.build();
    expect(output).toContain('Test List');
    expect(output).toContain('Item 1');
    expect(output).toContain('Item 2');
    // Check for boxen characters
    expect(output).toMatch(/╭.*╮/);
    expect(output).toMatch(/╰.*╯/);
  });

  it('should render a tree card with borders', () => {
    const builder = new VerboseCardBuilder();
    const treeContent: CardContent = {
      type: 'tree',
      root: {
        label: 'root',
        children: [
          { label: 'child1' },
          { label: 'child2' },
        ],
      },
    };
    builder.addCard('Test Tree', treeContent, 'warning');
    const output = builder.build();
    expect(output).toContain('Test Tree');
    expect(output).toContain('root');
    expect(output).toContain('child1');
    expect(output).toContain('child2');
    // Check for boxen characters
    expect(output).toMatch(/╭.*╮/);
    expect(output).toMatch(/╰.*╯/);
  });
});

describe('VerboseCardBuilder private methods', () => {
  it('getStyleOptions should return correct styles', () => {
    const builder = new VerboseCardBuilder();
    // @ts-ignore
    expect(builder.getStyleOptions('info')).toEqual({ borderStyle: 'round', borderColor: 'blue' });
    // @ts-ignore
    expect(builder.getStyleOptions('success')).toEqual({ borderStyle: 'round', borderColor: 'green' });
    // @ts-ignore
    expect(builder.getStyleOptions('warning')).toEqual({ borderStyle: 'round', borderColor: 'yellow' });
    // @ts-ignore
    expect(builder.getStyleOptions('error')).toEqual({ borderStyle: 'round', borderColor: 'red' });
    // @ts-ignore
    expect(builder.getStyleOptions('dim')).toEqual({ borderStyle: 'round', borderColor: 'gray', dimBorder: true });
    // @ts-ignore
    expect(builder.getStyleOptions('default')).toEqual({ borderStyle: 'round' });
  });

  it('getBoxenOptions should handle NO_COLOR', () => {
    const builder = new VerboseCardBuilder();
    process.env.NO_COLOR = 'true';
    // @ts-ignore
    const options = builder.getBoxenOptions('Test', 'info');
    expect(options.borderStyle).toBe('single');
    expect(options.borderColor).toBeUndefined();
    delete process.env.NO_COLOR;
  });
});

describe('VerboseCardBuilder snapshots', () => {
  const styles: CardStyle[] = ['info', 'success', 'warning', 'error', 'dim', 'default'];
  for (const style of styles) {
    it(`should match snapshot for ${style} style`, () => {
      const builder = new VerboseCardBuilder();
      builder.addCard(`Test ${style}`, 'This is a test card', style);
      const output = builder.build();
      expect(output).toMatchSnapshot();
    });
  }
});
