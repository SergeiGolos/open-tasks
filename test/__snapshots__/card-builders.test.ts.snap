// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VerboseCardBuilder snapshots > should match snapshot for default style 1`] = `
"â•­ Test default â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚                         â”‚
â”‚   This is a test card   â”‚
â”‚                         â”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯"
`;

exports[`VerboseCardBuilder snapshots > should match snapshot for dim style 1`] = `
"[2m[90mâ•­ Test dim â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[39m[22m
[2m[90mâ”‚[39m[22m                         [2m[90mâ”‚[39m[22m
[2m[90mâ”‚[39m[22m   This is a test card   [2m[90mâ”‚[39m[22m
[2m[90mâ”‚[39m[22m                         [2m[90mâ”‚[39m[22m
[2m[90mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[39m[22m"
`;

exports[`VerboseCardBuilder snapshots > should match snapshot for error style 1`] = `
"[31mâ•­ Test error â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[39m
[31mâ”‚[39m                         [31mâ”‚[39m
[31mâ”‚[39m   This is a test card   [31mâ”‚[39m
[31mâ”‚[39m                         [31mâ”‚[39m
[31mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[39m"
`;

exports[`VerboseCardBuilder snapshots > should match snapshot for info style 1`] = `
"[34mâ•­ Test info â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[39m
[34mâ”‚[39m                         [34mâ”‚[39m
[34mâ”‚[39m   This is a test card   [34mâ”‚[39m
[34mâ”‚[39m                         [34mâ”‚[39m
[34mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[39m"
`;

exports[`VerboseCardBuilder snapshots > should match snapshot for success style 1`] = `
"[32mâ•­ Test success â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[39m
[32mâ”‚[39m                         [32mâ”‚[39m
[32mâ”‚[39m   This is a test card   [32mâ”‚[39m
[32mâ”‚[39m                         [32mâ”‚[39m
[32mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[39m"
`;

exports[`VerboseCardBuilder snapshots > should match snapshot for warning style 1`] = `
"[33mâ•­ Test warning â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[39m
[33mâ”‚[39m                         [33mâ”‚[39m
[33mâ”‚[39m   This is a test card   [33mâ”‚[39m
[33mâ”‚[39m                         [33mâ”‚[39m
[33mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[39m"
`;
